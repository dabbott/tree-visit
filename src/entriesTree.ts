import { diagram } from './diagram'
import { find } from './find'
import { IndexPath } from './indexPath'

export type BaseEntriesOptions<T, PK extends PropertyKey> = {
  getEntries: (node: T, keyPath: PK[]) => [PK, T][]
  getChild?: (parent: T, parentKeyPath: PK[], childKey: PK) => T
}

export type FindEntriesOptions<T, PK extends PropertyKey> = {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, keyPath: PK[]) => boolean
}

export type DiagramEntriesOptions<T, PK extends PropertyKey> = {
  getLabel: (node: T, keyPath: PK[]) => string
}

// For each slice of the index path get the mapped key and create a new path from them
export function resolveIndexPath<PK extends PropertyKey>(
  map: Map<string, PK>,
  indexPath: IndexPath
): PK[] {
  const path: PK[] = []
  let current = ''

  for (let i = 0; i < indexPath.length; i++) {
    if (i > 0) current += '.'
    current += indexPath[i]
    const key = map.get(current)
    path.push(key!)
  }

  return path
}

function createEntriesAdapter<T, PK extends PropertyKey>(
  options: BaseEntriesOptions<T, PK>
) {
  let map = new Map<string, PK>()

  function getChildren([key, node]: [PK, T], indexPath: IndexPath) {
    map.set(indexPath.join('.'), key)

    const path = resolveIndexPath(map, indexPath)

    return options.getEntries(node, path)
  }

  function getNodePath(indexPath: IndexPath, key: PK) {
    return indexPath.length > 0
      ? [...resolveIndexPath(map, indexPath.slice(0, -1)), key]
      : []
  }

  return {
    getChildren,
    getNodePath,
  }
}

export function defineEntriesTree<T, PK extends PropertyKey>(
  options: BaseEntriesOptions<T, PK>
) {
  function entriesFind(node: T, findOptions: FindEntriesOptions<T, PK>) {
    const { getChildren, getNodePath } = createEntriesAdapter(options)

    const result = find(['' as PK, node], {
      getChildren,
      predicate: ([key, node], indexPath: IndexPath) => {
        return findOptions.predicate(node, getNodePath(indexPath, key))
      },
    })

    return result && result[1]
  }

  function entriesDiagram(
    node: T,
    diagramOptions: DiagramEntriesOptions<T, PK>
  ) {
    const { getChildren, getNodePath } = createEntriesAdapter(options)

    return diagram(['' as PK, node], {
      getChildren,
      getLabel: ([key, node], indexPath: IndexPath) => {
        return diagramOptions.getLabel(node, getNodePath(indexPath, key))
      },
    })
  }

  return { diagram: entriesDiagram, find: entriesFind }
}
