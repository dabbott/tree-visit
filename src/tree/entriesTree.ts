import { access, accessPath } from '../access'
import { DiagramEntriesOptions, diagram } from '../diagram'
import {
  FindEntriesOptions,
  FindEntriesOptionsTyped,
  find,
  findAll,
  findAllIndexPaths,
  findIndexPath,
} from '../find'
import { flat } from '../flat'
import { BaseEntriesOptions } from '../options'
import { ExtractRequiredKeys, OptionCheck, Prettify } from '../types'
import { VisitEntriesOptions, visit } from '../visit'

type WithoutBase<T, PK extends PropertyKey> = Omit<
  T,
  keyof BaseEntriesOptions<T, PK>
>

// type MutationOptions<T, PK extends PropertyKey> = WithoutBase<MutationBaseOptions<T>, PK>

type DiagramOptionsWB<T, PK extends PropertyKey> = WithoutBase<
  DiagramEntriesOptions<T, PK>,
  PK
>
type DiagramRequiredOptions<T, PK extends PropertyKey> = Pick<
  DiagramOptionsWB<T, PK>,
  ExtractRequiredKeys<DiagramOptionsWB<T, PK>>
>
type DiagramOptionalOptions<T, PK extends PropertyKey> = Omit<
  DiagramOptionsWB<T, PK>,
  ExtractRequiredKeys<DiagramOptionsWB<T, PK>>
>
type FindOptionsWB<T, PK extends PropertyKey> = WithoutBase<
  FindEntriesOptions<T, PK>,
  PK
>
type VisitOptionsWB<T, PK extends PropertyKey> = WithoutBase<
  VisitEntriesOptions<T, PK>,
  PK
>
// type InsertOptionsWB<T> = WithoutBase<InsertOptions<T>>
// type RemoveOptionsWB<T> = WithoutBase<RemoveOptions<T>>
// type MoveOptionsWB<T> = WithoutBase<MoveOptions<T>>
// type ReplaceOptionsWB<T> = WithoutBase<ReplaceOptions<T>>

type ApplyableOptions<T, PK extends PropertyKey> = DiagramRequiredOptions<T, PK> // & MutationOptions<T>

interface Overloads<T, PK extends PropertyKey> {
  /**
   * Find a node matching a predicate function.
   */
  find(
    node: T,
    predicate: FindEntriesOptions<T, PK>['predicate']
  ): T | undefined

  find(node: T, options: FindOptionsWB<T, PK>): T | undefined

  find<S extends T>(
    node: T,
    predicate: FindEntriesOptionsTyped<T, PK, S>['predicate']
  ): S | undefined

  find<S extends T>(
    node: T,
    options: WithoutBase<FindEntriesOptionsTyped<T, PK, S>, PK>
  ): S | undefined

  /**
   * Find all nodes matching a predicate function.
   */
  findAll(node: T, predicate: FindEntriesOptions<T, PK>['predicate']): T[]

  findAll(node: T, options: FindOptionsWB<T, PK>): T[]

  findAll<S extends T>(
    node: T,
    predicate: FindEntriesOptionsTyped<T, PK, S>['predicate']
  ): S[]

  findAll<S extends T>(
    node: T,
    options: WithoutBase<FindEntriesOptionsTyped<T, PK, S>, PK>
  ): S[]

  /**
   * Find the `IndexPath` of a node matching a predicate function.
   */
  findIndexPath(
    node: T,
    predicate: FindEntriesOptions<T, PK>['predicate']
  ): PK[] | undefined

  findIndexPath(node: T, options: FindOptionsWB<T, PK>): PK[] | undefined

  /**
   * Find the `IndexPath` of all nodes matching a predicate function.
   */
  findAllIndexPaths(
    node: T,
    predicate: FindEntriesOptions<T, PK>['predicate']
  ): PK[][]

  findAllIndexPaths(node: T, options: FindOptionsWB<T, PK>): PK[][]

  /**
   * Visit each node using preorder DFS traversal.
   */
  visit(
    node: T,
    onEnter: NonNullable<VisitEntriesOptions<T, PK>>['onEnter']
  ): void

  /**
   * Visit each node using DFS traversal.
   */
  visit(node: T, options: VisitOptionsWB<T, PK>): void
}

export class EntriesTree<
  T,
  PK extends PropertyKey,
  AppliedOptions extends Partial<ApplyableOptions<T, PK>>
> {
  constructor(
    private baseOptions: BaseEntriesOptions<T, PK>,
    private appliedOptions: AppliedOptions
  ) {}

  private mergeOptions = <T extends Record<string, any>>(options: T) => ({
    ...this.baseOptions,
    ...this.appliedOptions,
    ...options,
  })

  withOptions = <NewOptions extends Partial<ApplyableOptions<T, PK>>>(
    newOptions: NewOptions
  ) =>
    new EntriesTree(this.baseOptions, {
      ...this.appliedOptions,
      ...newOptions,
    })

  /**
   * Returns a node by its `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  access = (node: T, path: PK[]) => access(node, this.mergeOptions({ path }))

  /**
   * Returns an array of each node in an `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  accessPath = (node: T, path: PK[]) =>
    accessPath(node, this.mergeOptions({ path }))

  /**
   * Generate a diagram of the tree, as a string.
   */
  diagram = (
    node: T,
    options:
      | DiagramRequiredOptions<T, PK>['getLabel']
      | Prettify<
          AppliedOptions extends DiagramRequiredOptions<T, PK>
            ? DiagramOptionalOptions<T, PK> | void
            : OptionCheck<
                AppliedOptions,
                'getLabel',
                DiagramRequiredOptions<T, PK>
              > &
                DiagramOptionalOptions<T, PK>
        >
  ) =>
    typeof options === 'function'
      ? diagram(node, this.mergeOptions({ getLabel: options }))
      : diagram(node, this.mergeOptions(options))

  find: Overloads<T, PK>['find'] = (
    node: T,
    predicateOrOptions:
      | FindEntriesOptions<T, PK>['predicate']
      | FindOptionsWB<T, PK>
  ) =>
    typeof predicateOrOptions === 'function'
      ? find(node, this.mergeOptions({ predicate: predicateOrOptions }))
      : find(node, this.mergeOptions({ ...predicateOrOptions }))

  findAll: Overloads<T, PK>['findAll'] = (
    node: T,
    predicateOrOptions:
      | FindEntriesOptions<T, PK>['predicate']
      | FindOptionsWB<T, PK>
  ) =>
    typeof predicateOrOptions === 'function'
      ? findAll(node, this.mergeOptions({ predicate: predicateOrOptions }))
      : findAll(node, this.mergeOptions({ ...predicateOrOptions }))

  findIndexPath: Overloads<T, PK>['findIndexPath'] = (
    node: T,
    predicateOrOptions:
      | FindEntriesOptions<T, PK>['predicate']
      | FindOptionsWB<T, PK>
  ) =>
    typeof predicateOrOptions === 'function'
      ? findIndexPath(
          node,
          this.mergeOptions({ predicate: predicateOrOptions })
        )
      : findIndexPath(node, this.mergeOptions({ ...predicateOrOptions }))

  findAllIndexPaths: Overloads<T, PK>['findAllIndexPaths'] = (
    node: T,
    predicateOrOptions:
      | FindEntriesOptions<T, PK>['predicate']
      | FindOptionsWB<T, PK>
  ) =>
    typeof predicateOrOptions === 'function'
      ? findAllIndexPaths(
          node,
          this.mergeOptions({ predicate: predicateOrOptions })
        )
      : findAllIndexPaths(node, this.mergeOptions({ ...predicateOrOptions }))

  /**
   * Returns an array containing the root node and all of its descendants.
   *
   * This is analogous to `Array.prototype.flat` for flattening arrays.
   */
  flat = (node: T) => flat(node, this.mergeOptions({}))

  // /**
  //  * Map each node into an array of values, which are then flattened into a single array.
  //  *
  //  * This is analogous to `Array.prototype.flatMap` for arrays.
  //  */
  // flatMap = <R>(node: T, transform: FlatMapOptions<T, R>['transform']) =>
  //   flatMap(node, this.mergeOptions({ transform }))

  // reduce = <R>(
  //   node: T,
  //   nextResult: ReduceChildrenOptions<T, R>['nextResult'],
  //   initialResult: R
  // ): R => reduce(node, this.mergeOptions({ nextResult, initialResult }))

  // map = <R>(node: T, transform: MapOptions<T, R>['transform']): R =>
  //   map(node, this.mergeOptions({ transform }))

  visit: Overloads<T, PK>['visit'] = (
    node: T,
    onEnterOrOptions:
      | NonNullable<VisitEntriesOptions<T, PK>>['onEnter']
      | VisitOptionsWB<T, PK>
  ) =>
    typeof onEnterOrOptions === 'function'
      ? visit(node, this.mergeOptions({ onEnter: onEnterOrOptions }))
      : visit(node, this.mergeOptions({ ...onEnterOrOptions }))

  // --- Mutations ---

  /**
   * Insert nodes at a given `IndexPath`.
   */
  // insert = (
  //   node: T,
  //   options: Prettify<
  //     OptionCheck<AppliedOptions, 'create', InsertOptionsWB<T>> &
  //       Omit<InsertOptionsWB<T>, 'create'>
  //   >
  // ) => insert(node, this.mergeOptions(options))

  // /**
  //  * Remove nodes at the given `IndexPath`s.
  //  */
  // remove = (
  //   node: T,
  //   options: Prettify<
  //     OptionCheck<AppliedOptions, 'create', RemoveOptionsWB<T>> &
  //       Omit<RemoveOptionsWB<T>, 'create'>
  //   >
  // ) => remove(node, this.mergeOptions(options))

  // /**
  //  * Move nodes from one `IndexPath` to another.
  //  */
  // move = (
  //   node: T,
  //   options: Prettify<
  //     OptionCheck<AppliedOptions, 'create', MoveOptionsWB<T>> &
  //       Omit<MoveOptionsWB<T>, 'create'>
  //   >
  // ) => move(node, this.mergeOptions(options))

  // /**
  //  * Replace the node at the given `IndexPath` with another
  //  */
  // replace = (
  //   node: T,
  //   options: Prettify<
  //     OptionCheck<AppliedOptions, 'create', ReplaceOptionsWB<T>> &
  //       Omit<ReplaceOptionsWB<T>, 'create'>
  //   >
  // ) => replace(node, this.mergeOptions(options))
}
