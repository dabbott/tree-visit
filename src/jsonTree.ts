import { defineEntriesTree } from './entriesTree'

export type JSONValue =
  | string
  | number
  | boolean
  | null
  | JSONValue[]
  | { [key: string]: JSONValue }

export const JSONTree = defineEntriesTree({
  getEntries(node: JSONValue) {
    if (typeof node === 'object' && node !== null) {
      return Object.entries(node)
    } else {
      return []
    }
  },
})
