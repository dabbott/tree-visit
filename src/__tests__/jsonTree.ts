import { resolveIndexPath } from '../entriesTree'
import { JSONTree, JSONValue } from '../jsonTree'

it('resolves index path', () => {
  const map = new Map<string, string>([
    ['0', 'foo'],
    ['0.0', 'bar'],
    ['0.0.0', 'baz'],
    ['0.0.1', 'qux'],
    ['0.1', 'quux'],
    ['1', 'corge'],
  ])

  expect(resolveIndexPath(map, [])).toEqual([])
  expect(resolveIndexPath(map, [0])).toEqual(['foo'])
  expect(resolveIndexPath(map, [0, 0])).toEqual(['foo', 'bar'])
  expect(resolveIndexPath(map, [0, 0, 0])).toEqual(['foo', 'bar', 'baz'])
  expect(resolveIndexPath(map, [0, 0, 1])).toEqual(['foo', 'bar', 'qux'])
  expect(resolveIndexPath(map, [0, 1])).toEqual(['foo', 'quux'])
  expect(resolveIndexPath(map, [1])).toEqual(['corge'])
})

it('displays json tree diagram', () => {
  expect(
    JSONTree.diagram(
      {
        null: null,
        boolean: true,
        number: 42,
        string: 'hello',
        multilineString: 'line1\nline2',
        array: [1, 2, 3],
        object: { foo: 'bar', baz: { qux: 'quux' } },
      },
      {
        getLabel,
      }
    )
  ).toMatchSnapshot()
})

it('finds node', () => {
  const node = {
    foo: {
      bar: {
        baz: 'hello',
      },
    },
  }

  const result = JSONTree.find(node, {
    predicate: (node, path) => {
      return path.join('.') === 'foo.bar'
    },
  })

  expect(result).toEqual({ baz: 'hello' })
})

function getLabel(node: JSONValue, keyPath: string[]) {
  // console.log({ node, keyPath })
  const key = keyPath[keyPath.length - 1]
  const text = printJSONValue(node)
  return `${key === undefined ? text : key + ': ' + text} [${keyPath.join(
    '/'
  )}]`
}

function printJSONValue(value: JSONValue): string {
  if (typeof value === 'object' && value !== null) {
    return `<${Object.prototype.toString.call(value).slice(8, -1)}>`
  } else {
    return JSON.stringify(value)
  }
}
