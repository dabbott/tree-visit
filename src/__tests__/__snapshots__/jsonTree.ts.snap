// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays json tree diagram 1`] = `
"<Object> []
├── null: null [null]
├── boolean: true [boolean]
├── number: 42 [number]
├── string: \\"hello\\" [string]
├── multilineString: \\"line1\\\\nline2\\" [multilineString]
├── array: <Array> [array]
│   ├── 0: 1 [array/0]
│   ├── 1: 2 [array/1]
│   └── 2: 3 [array/2]
└── object: <Object> [object]
    ├── foo: \\"bar\\" [object/foo]
    └── baz: <Object> [object/baz]
        └── qux: \\"quux\\" [object/baz/qux]"
`;
